SDWebImage/SDMemoryNSCache.m

```diff
@@ -51,17 +50,16 @@ - (instancetype)initWithConfig:(SDImageCacheConfig *)config {
 }
 
 - (void)commonInit {
-    self.weakCache = [[NSMapTable alloc] initWithKeyOptions:NSPointerFunctionsStrongMemory valueOptions:NSPointerFunctionsWeakMemory capacity:0];
-    self.weakCacheLock = dispatch_semaphore_create(1);
-    
     SDImageCacheConfig *config = self.config;
     self.totalCostLimit = config.maxMemoryCost;
     self.countLimit = config.maxMemoryCount;
     
-    [config addObserver:self forKeyPath:NSStringFromSelector(@selector(maxMemoryCost)) options:0 context:SDMemoryCacheContext];
-    [config addObserver:self forKeyPath:NSStringFromSelector(@selector(maxMemoryCount)) options:0 context:SDMemoryCacheContext];
+    [config addObserver:self forKeyPath:NSStringFromSelector(@selector(maxMemoryCost)) options:0 context:SDMemoryNSCacheContext];
+    [config addObserver:self forKeyPath:NSStringFromSelector(@selector(maxMemoryCount)) options:0 context:SDMemoryNSCacheContext];
     
 #if SD_UIKIT
+    self.weakCache = [[NSMapTable alloc] initWithKeyOptions:NSPointerFunctionsStrongMemory valueOptions:NSPointerFunctionsWeakMemory capacity:0];
+    self.weakCacheLock = dispatch_semaphore_create(1);
```

Comment:
`weakCache` just use for UIKit.

Reply:
So we just move back these code into the `SDMemoryCache` implementation. Don't change them. Let the weak cache works for both LRU/NSCache.

Reply:
Good idea.

Reply:
> @Insofan I changed `SDMemoryCache` class inherit `NSCache` to `NSObject`, From your code, There isn't necessary to inherit `NSCache` class and add KVO.
Inherit from NSObject will break SDMemoryCache api totalCostLimit & countLimit.

Reply:
Really？From your code, `SDMemoryCache` class never use property from `NSCache` :
```Ruby
@property NSUInteger totalCostLimit;	// limits are imprecise/not strict
@property NSUInteger countLimit;	// limits are imprecise/not strict
```
So This just for `NSCache` class , Not have API breaking.

Reply:
> @Insofan Hi. I think again about that weak cache. That is used to solve the problem:
> 
> > when memory warning occur, the cache was purged, but actually, some imageViews.image still keep a strong reference to image. So if user scroll and back, we can use weak cache to provide the image instance, to avoid disk cache query and cause a blank flash.

So weakCache in the `SDMemoryCache` class, actually I wanna  make the`SDMemoryCache` to be a absolutly abstract class, it will be easy to expand other memory strategies.

Reply:
@Insofan You also should keep `weak cache` for LRUCache.  and move back `weak cache` code into the `SDMemoryCache` implementation.

Reply:
> LGTM for my first site of review.
> 
> I'm a little busy now. I'll try to review again and maybe some patch for this. Don't be worried, this feature is added to 5.1.0 milestone, I think it's a great feature till now from 5.0.0-5.1.0.
> 
> If you don't have enough time to polish, I can merge this later, and create another PR for some no-trival changes (such as the Documentations, Test Cases, little Code Refactory, etc)

After the main logic be approve, I'll write some test caces and documents. I think performance testing should be done by someone else.

Reply:
@kinarob The test case  `test46MemoryCacheWeakCache`  in `SDImageCacheTest.m`  would be failed.

Reply:
Emm. I noticed

Reply:
@Infofan Actually...

The 5.0 cache system and image loader system, already been generatic and can be customized. See [Wiki- Memory Cache](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#memory-cache)

You can just use `SDImageCacheConfig.memoryCacheClass = MyClass`.

So that ` Actually I wanna make theSDMemoryCache to be a absolutly abstract class, I prefer generic code then if else statement, it will be easy to expand other memory strategies. And this abstract class structure could be used on disk cache.`. You can even don't add anything like `shouldUseLRUCache` this into the new property.

Just give them two cache class, `SDMemoryNSCache` and `SDMemoryLRUCache`, they set this into the memoryCacheClass. It's OK.

The only thing that I have to do so, it's the API break and semversion (You can not delete this `SDMemoryCache` class).

Current code using the `[self.memCache objectForKey:key]` is totally OK.

Reply:
For weak cache logic here, I think that you just leave those code. (Revert back). If you think it's a little tricky, I'll provide a refactory to use a more good code structure. But that is not related to your LRU implementation and this PR. We prefer to seperate different things in different PRs to simplify the code review and merge conflict thing.

Reply:
@Insofan It's helpful to add a friend with WeChat 😁😁.

Reply:
@dreampiggy Accroding to your WeChat message log, rollback the `SDMemoryCache` class code, delete the `SDMemoryNSCache` class, just add a `SDMemoryLRUCache` and a `shouldUseLRUCache` property in `SDImageCacheConfig`.How about this logic.

Reply:
I think this `shouldUseLRUCache` is useless. Why not just this ?

```SDImageCache.defaultCacheConfig.memoryCacheClass = SDMemoryLRUCache.class;```

Reply:
@dreampiggy I'm confusing about it. Emm.So what's the logic in `- (void)setObject:forKey:` method.

Reply:
@kinarob Do you have any suggestions for upper comment.

Reply:
Don't change any code of `SDMemoryCache`, If someone want use `SDMemoryLRUCache`, Like this:
```ruby
SDImageCache.defaultCacheConfig.memoryCacheClass = SDMemoryLRUCache.class;
```

Reply:
My worry is that `shouldUseLRUCache` property in `SDImageCacheConfig`. I'm agree with that we can keep only two class:

+ SDMemoryCache, which is subclass of `NSCache` and implements `SDMemoryCacheProtocol`
+ SDMemoryLRUCache, which is subclass of `NSObject` and implemnts `SDMemoryCacheProtocol`

They are in the same level, there are no another `SDMemoryNSCache` class. This design, it's because we encourage the combination but not inherency. Each implementation for protocol is seperated. There are no need for another abstract `SDMemoryCacheBaseClass` this design pattern.

Sorry for my previous code review that lead you some misunderstanding. At that time I'm a little hurry and forget the design of `SDMemoryCacheProtocol`, so I think we need a Abstact class design pattern for this problem. However, in the last weekend I realize my problem and provide the comment again for changes.

Reply:
@dreampiggy Request code review, I have session with kinarobin about SD_UIKIT part notification.Do you have any other advice?